drivers/platform/xilinx/spi.c

```diff
@@ -54,166 +54,235 @@
-			goto ps_error;
-
-		ret = XSpiPs_CfgInitialize(xdesc->instance,
-					   xdesc->config,
-					   ((XSpiPs_Config*)xdesc->config)
-					   ->BaseAddress);
-		if(ret != SUCCESS)
-			goto ps_error;
+		// this exponent - 1 is needed because of the way
+		// xilinx stores it into registers
+		if (prescaler)
+			prescaler -= 1;
 
-		if (sdesc->max_speed_hz != 0u) {
-			uint32_t div = input_clock / sdesc->max_speed_hz;
-			uint32_t rem = input_clock % sdesc->max_speed_hz;
+		// this exponent - 1 is needed when initial div was
+		// precisely a power of two
+		if (!rem && prescaler)
+			prescaler -= 1;
 
-			// find the power of two just higher than div and
-			// store the exponent in prescaler
-			while(div) {
-				prescaler += 1;
-				div >>= 1u;
-			}
+		if (prescaler < prescaler_min)
+			prescaler = prescaler_min;
 
-			// this exponent - 1 is needed because of the way
-			// xilinx stores it into registers
-			if (prescaler)
-				prescaler -= 1;
+		if (prescaler > prescaler_max)
+			prescaler = prescaler_max;
+	} else
+		prescaler = prescaler_default;
 
-			// this exponent - 1 is needed when initial div was
-			// precisely a power of two
-			if (!rem && prescaler)
-				prescaler -= 1;
+	ret = XSpiPs_SetClkPrescaler(xdesc->instance, prescaler);
 
-			if (prescaler < prescaler_min)
-				prescaler = prescaler_min;
+	if(ret != SUCCESS)
+		goto ps_error;
 
-			if (prescaler > prescaler_max)
-				prescaler = prescaler_max;
-		} else
-			prescaler = prescaler_default;
+	return SUCCESS;
 
-		ret = XSpiPs_SetClkPrescaler(xdesc->instance, prescaler);
+ps_error:
+	free(xdesc->instance);
+	free(xdesc);
+#endif
+	return FAILURE;
+}
 
-		if(ret != SUCCESS)
-			goto ps_error;
+/**
+ * @brief Initialize the SPI communication peripheral.
+ * @param desc - The SPI descriptor.
+ * @param param - The structure that contains the SPI parameters.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t spi_init(struct spi_desc **desc,
+		 const struct spi_init_param *param)
+{
+	int32_t				ret;
+	enum xil_spi_type		*spi_type;
+
+	if (!param) {
+		return FAILURE;
+	}
+
+	*desc = malloc(sizeof(**desc));
+	if(!*desc) {
+		free(*desc);
+		return FAILURE;
+	}
+
+	/* Get only the first member of the sctructure
+	 * Both structures (spi_engine_init_param and spi_init_param)
+	 * have the fisrt member enum xil_spi_type so in this case
+	 * we can read the first member without casting the pointer to
+	 * a certain structure type.
+	*/
+	spi_type = param->extra;
+
+	if(!spi_type)
+		goto init_error;
 
+	(*desc)->max_speed_hz = param->max_speed_hz;
+	(*desc)->mode = param->mode;
+	(*desc)->chip_select = param->chip_select;
+
+	switch (*spi_type) {
+	case SPI_PL:
+		ret = spi_init_pl(desc, param);
+		break;
+	case SPI_PS:
+		ret = spi_init_ps(desc, param);
 		break;
-ps_error:
-		free(xdesc->instance);
-#endif
-		goto error;
 	case SPI_ENGINE:
 #ifdef SPI_ENGINE_H
-		//TODO: Implement SPI engine feature
-
+		ret = spi_engine_init(desc, param);
```

Comment:
makes more sense if the header will be 
`int32_t spi_engine_init(struct spi_desc *desc, const struct spi_init_param *param)`
because the general descriptor is not allocated inside of these inits.
And the call to this functions will be:
```
ret = spi_init_pl(*desc, param);
ret = spi_init_ps(*desc, param);
ret = spi_engine_init(*desc, param);
```